error CS1056: Unexpected character '−' with Regex on Mac

### Description

Compilation error CS1056 on Mac when using `GeneratedRegex` with this regular expression:

```
[GeneratedRegex("(?>'[^']*'|\"[^\"]*\"|\\([^()]*(((?<open>\\()[^()]*)+((?<-open>\\))[^()]*)+)*(?(open)(?!))\\)|[^'\",])+", RegexOptions.IgnoreCase | RegexOptions.Compiled, "en")]
```

### Reproduction Steps

On MacOS (Sonoma 14.4) with dotnet 9.0.200, create a minimal project and compile with `dotnet build`:

```csharp
using System.Text.RegularExpressions;

namespace ConsoleApp1
{
    internal partial class Program
    {
        [GeneratedRegex("(?>'[^']*'|\"[^\"]*\"|\\([^()]*(((?<open>\\()[^()]*)+((?<-open>\\))[^()]*)+)*(?(open)(?!))\\)|[^'\",])+", RegexOptions.IgnoreCase | RegexOptions.Compiled, "en")]
        private static partial Regex FindColumnsPattern();
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
        }
    }
}
```

### Expected behavior

The program compiles

### Actual behavior

The program does not compile and got this error

```
ConsoleApp1 failed with 1 error(s) (0,1s)
    /Users/carlos.saito/tmp/weird-regex/ConsoleApp1/obj/Debug/net9.0/System.Text.RegularExpressions.Generator/System.Text.RegularExpressions.Generator.RegexGenerator/RegexGenerator.g.cs(409,78): error CS1056: Unexpected character '−'
```

Note: in the console message, the `−` in _Unexpected character '−'_ is a [minus sign](https://www.compart.com/en/unicode/U+2212) which is not present in the source code

### Regression?

_No response_

### Known Workarounds

Removing the `-` in the regex capture group compiles, but it is a different expression:

```diff
- [GeneratedRegex("(?>'[^']*'|\"[^\"]*\"|\\([^()]*(((?<open>\\()[^()]*)+((?<-open>\\))[^()]*)+)*(?(open)(?!))\\)|[^'\",])+", RegexOptions.IgnoreCase | RegexOptions.Compiled, "en")]
+ [GeneratedRegex("(?>'[^']*'|\"[^\"]*\"|\\([^()]*(((?<open>\\()[^()]*)+((?<open>\\))[^()]*)+)*(?(open)(?!))\\)|[^'\",])+", RegexOptions.IgnoreCase | RegexOptions.Compiled, "en")]
                                                                            ^ here
```

### Configuration

```
$ dotnet --version
9.0.200
```

```
$ dotnet sdk check
.NET SDKs:
Version      Status     
------------------------
9.0.200      Up to date.

.NET Runtimes:
Name                          Version      Status     
------------------------------------------------------
Microsoft.AspNetCore.App      9.0.2        Up to date.
Microsoft.NETCore.App         9.0.2        Up to date.
```

Running on Mac OS 14.4

### Other information

_No response_

--- a/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs
+++ b/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs
@@ -2573,7 +2573,7 @@ namespace System.Text.RegularExpressions.Generator
                     }
 
                     writer.WriteLine();
-                    writer.WriteLine($"base.TransferCapture({capnum}, {uncapnum}, {startingPos}, pos);");
+                    writer.WriteLine($"base.TransferCapture({capnum.ToString(CultureInfo.InvariantCulture)}, {uncapnum}, {startingPos}, pos);");
                 }
 
                 if (isAtomic || !childBacktracks)

--- a/src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexGeneratorHelper.netcoreapp.cs
+++ b/src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexGeneratorHelper.netcoreapp.cs
@@ -134,6 +134,12 @@ namespace System.Text.RegularExpressions.Tests
             return results[0];
         }
 
+        private static readonly CultureInfo s_cultureWithMinusNegativeSign = new CultureInfo("")
+        {
+            // To validate that generation still succeeds even when something other than '-' is used.
+            NumberFormat = new NumberFormatInfo() { NegativeSign = $"{(char)0x2212}" }
+        };
+
         internal static async Task<Regex[]> SourceGenRegexAsync(
             (string pattern, CultureInfo? culture, RegexOptions? options, TimeSpan? matchTimeout)[] regexes, CancellationToken cancellationToken = default)
         {
@@ -214,13 +220,24 @@ namespace System.Text.RegularExpressions.Tests
             comp = comp.ReplaceSyntaxTree(comp.SyntaxTrees.First(), CSharpSyntaxTree.ParseText(SourceText.From(code.ToString(), Encoding.UTF8), s_previewParseOptions));
 
             // Run the generator
-            GeneratorDriverRunResult generatorResults = s_generatorDriver.RunGenerators(comp!, cancellationToken).GetRunResult();
-            ImmutableArray<Diagnostic> generatorDiagnostics = generatorResults.Diagnostics.RemoveAll(d => d.Severity <= DiagnosticSeverity.Hidden);
-            if (generatorDiagnostics.Length != 0)
+            CultureInfo origCulture = CultureInfo.CurrentCulture;
+            CultureInfo.CurrentCulture = s_cultureWithMinusNegativeSign;
+            GeneratorDriverRunResult generatorResults;
+            ImmutableArray<Diagnostic> generatorDiagnostics;
+            try
             {
-                throw new ArgumentException(
-                    string.Join(Environment.NewLine, generatorResults.GeneratedTrees.Select(t => NumberLines(t.ToString()))) + Environment.NewLine +
-                    string.Join(Environment.NewLine, generatorDiagnostics));
+                generatorResults = s_generatorDriver.RunGenerators(comp!, cancellationToken).GetRunResult();
+                generatorDiagnostics = generatorResults.Diagnostics.RemoveAll(d => d.Severity <= DiagnosticSeverity.Hidden);
+                if (generatorDiagnostics.Length != 0)
+                {
+                    throw new ArgumentException(
+                        string.Join(Environment.NewLine, generatorResults.GeneratedTrees.Select(t => NumberLines(t.ToString()))) + Environment.NewLine +
+                        string.Join(Environment.NewLine, generatorDiagnostics));
+                }
+            }
+            finally
+            {
+                CultureInfo.CurrentCulture = origCulture;
             }
 
             // Compile the assembly to a stream

